src/main/java/dataMgmt/SessionStorage.java

```diff
@@ -16,322 +19,507 @@
+	 * @param fileID
+	 *            The ID of the file to map the path to
+	 */
+	public void putAbsoluteFilePath(Path absPath, long projectID, long fileID) {
+		synchronized (FILE_LIST) {
+			Project p = projects.get(projectID);
+			p.putAbsoluteFilePath(absPath, fileID);
+		}
+	}
+
+	/**
+	 * Looks through all projects for the file that matches the given file ID
+	 * and returns it if found. If not found, null is returned.
+	 * 
+	 * @return The removed file
+	 */
+	public File getFile(long fileID) {
+		for (Project p : projects.values()) {
+			File f = p.getFile(fileID);
+			if (f != null) {
+				return f;
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * Looks through all projects for the file that matches the given absolute
+	 * path and returns it if found. If not found, this could either mean that
+	 * the file does not exist or there is no path mapped to the file.
+	 * 
+	 * @param absolutePath
+	 *            The absolute path of the file
+	 * @return The file that the path is mapped to, or null if the path is not
+	 *         mapped to a file
+	 */
+	public File getFile(Path absolutePath) {
+		for (Project p : projects.values()) {
+			File f = p.getFile(absolutePath);
+			if (f != null) {
+				return f;
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * Looks through all projects for the file that matches the given file ID
+	 * and removes it. If found, the file object is returned, otherwise this
+	 * method returns null.
+	 * 
+	 * @param fileID
+	 *            The ID of the file to remove
+	 * @return The file that is removed or null if there is no file with this ID
+	 *         in the store
+	 */
+	public File removeFile(long fileID) {
+		synchronized (FILE_LIST) {
+			for (Project p : projects.values()) {
+				File old;
+				if ((old = p.removeFile(fileID)) != null) {
+					return old;
+				}
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * Set the given project id as a subscribed project, if it exists.
+	 * 
+	 * @param id
+	 *            to set subscribed
+	 */
+	public void setSubscribed(long id) {
+		boolean notify = false;
+		synchronized (SUBSCRIBED_PROJECTS) {
+			if (this.projects.containsKey(id)) {
+				this.subscribedIds.add(id);
+				notify = true;
+			}
+		}
+		if (notify) {
+			notifyListeners(SUBSCRIBED_PROJECTS, null, id);
+		}
+	}
+
+	/**
+	 * Remove the given project id from the set of subscribed ids, if it exists.
+	 * 
+	 * @param id
+	 *            to remove from subscribed set
+	 */
+	public void setUnsubscribed(long id) {
+		boolean notify = false;
+		synchronized (SUBSCRIBED_PROJECTS) {
+			if (this.projects.containsKey(id)) {
+				this.subscribedIds.remove(id);
+				notify = true;
+			}
+		}
+		if (notify) {
+			notifyListeners(SUBSCRIBED_PROJECTS, id, null);
+		}
+	}
+
+	/**
+	 * Get the set of subscribed ids
+	 * 
+	 * @return set of subscribed ids
+	 */
+	public Set<Long> getSubscribedIds() {
+		Set<Long> ids;
+		synchronized (SUBSCRIBED_PROJECTS) {
+			ids = this.subscribedIds;
```

Comment:
This function can just be synchronized `1(SUBSCRIBED_PROJECTS) { return Collections.unmodifiableSet(subscribedIds); }` . Confirmed with @ObsessiveOrange. Notice you should make this an unmodifiable set as well.